{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "93758227-864d-4a6e-a697-a0a6562b89ff",
   "metadata": {},
   "source": [
    "The script needs to be run under linux mode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "delayed-sandwich",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "pd.set_option('display.max_rows', 320)\n",
    "\n",
    "from Bio import Seq\n",
    "from Bio.Seq import Seq\n",
    "from Bio import Restriction as Res\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "import math\n",
    "import itertools\n",
    "import time\n",
    "import nupack as npk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "objective-bronze",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://www.genscript.com/tools/codon-frequency-table\n",
    "\n",
    "# Dictionary containing sense codons for amino acids. For each aa, first list\n",
    "# contains cumulative probability of each codon occurence. Second list contains\n",
    "# all possible codons for each aa. Third list only contains codons which occur at probability of 0.2 or greater. \n",
    "# The fourth list is the corresponding filtered codons.\n",
    "sense_5to3_stats = {\n",
    "    'R': ([.08, .11, .18, .20, .21, .21],\n",
    "          ['cgt', 'cga', 'cgc', 'cgg', 'agg', 'aga'],\n",
    "          [.21, .21, .20],\n",
    "          ['aga', 'agg','cgg']), \n",
    "    'H': ([.42, .58],\n",
    "          ['cat', 'cac'], \n",
    "          [.58, .42],\n",
    "          ['cac', 'cat']),\n",
    "    'K': ([.43, .57],\n",
    "          ['aaa', 'aag'],\n",
    "          [.57, .43],\n",
    "          ['aag', 'aaa']),\n",
    "    'D': ([.46, .54],\n",
    "          ['gat', 'gac'],\n",
    "          [.54, .46],\n",
    "          ['gac', 'gat']),\n",
    "    'E': ([.42, .58],\n",
    "          ['gaa', 'gag'],\n",
    "          [.58, .42],\n",
    "          ['gag', 'gaa']),\n",
    "    'S': ([.05, .15, .15, .19, .22, .24],\n",
    "          ['tcg', 'tca', 'agt', 'tct', 'tcc', 'agc'],\n",
    "          [.24, .22],\n",
    "          ['agc', 'tcc']),\n",
    "    'T': ([.25, .28, .36],\n",
    "          ['act', 'aca', 'acc'],\n",
    "          [.36, .28, .25],\n",
    "          ['acc', 'aca', 'act']),\n",
    "    'N': ([.47, .53],\n",
    "          ['aat', 'aac'],\n",
    "          [.53, .47],\n",
    "          ['aac', 'aat']),\n",
    "    'Q': ([.27, .73],\n",
    "          ['caa', 'cag'],\n",
    "          [.73, .27],\n",
    "          ['cag', 'caa']),\n",
    "    'C': ([.46, .54],\n",
    "          ['tgt', 'tgc'],\n",
    "          [.54, .46],\n",
    "          ['tgc', 'tgt']),\n",
    "    'G': ([.16, .25, .25, .34],\n",
    "          ['ggt', 'gga', 'ggg', 'ggc'],\n",
    "          [.34, .25, .25],\n",
    "          ['ggc', 'gga', 'ggg']),\n",
    "    'P': ([.11, .28, .29, .32],\n",
    "          ['ccg', 'cca', 'cct', 'ccc'],\n",
    "          [.32, .29, .28],\n",
    "          ['ccc', 'cct', 'cca']),\n",
    "    'A': ([.11, .23, .27, .4],\n",
    "          ['gcg', 'gca', 'gct', 'gcc'],\n",
    "          [.4, .27, .23],\n",
    "          ['gcc', 'gct', 'gca']),\n",
    "    'I': ([.17, .36, .47],\n",
    "          ['ata', 'att', 'atc'],\n",
    "          [.47, .36],\n",
    "          ['atc', 'att']),\n",
    "    'L': ([.07, .08, .13, .13, .2, .4],\n",
    "          ['cta', 'tta', 'ttg', 'ctt', 'ctc', 'ctg'],\n",
    "          [.4, .2],\n",
    "          ['ctg', 'ctc']),\n",
    "    'M': ([1],\n",
    "          ['atg'],\n",
    "          [1],\n",
    "          ['atg']),\n",
    "    'F': ([.46, .54],\n",
    "          ['ttt', 'ttc'],\n",
    "          [.54, .46],\n",
    "          ['ttc', 'ttt']),\n",
    "    'W': ([1],\n",
    "          ['tgg'],\n",
    "          [1],\n",
    "          ['tgg']),\n",
    "    'Y': ([.44, .56],\n",
    "          ['tat', 'tac'],\n",
    "          [.56, .44],\n",
    "          ['tac', 'tat']),\n",
    "    'V': ([.12, .18, .24, .46],\n",
    "          ['gta', 'gtt', 'gtc', 'gtg'],\n",
    "          [.46, .24],\n",
    "          ['gtg', 'gtc'])}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "driven-trance",
   "metadata": {},
   "source": [
    "## Input for the list of antigens"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "188a977d-707f-4c79-863d-76a14883ea74",
   "metadata": {},
   "source": [
    "The input file should be an .xlsx Excel spreadsheet with the following columns:\n",
    "1. the amino acid sequences of the antigens in single letter format, e.g. \"PSPSMGRDIKVQFQS\"\n",
    "2. optional information each stored in a column, such as \"Allele\", \"Disease\", \"Protein\", \"Position on the protein\" etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2631e433-d7e0-422b-9263-1a78c8f4d419",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_path = input(\"Please enter the file path for the antigen list: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "smart-butler",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## Get the peptide list\n",
    "oriFile = pd.DataFrame(pd.read_excel(file_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cathedral-horizon",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# clean up file and only show info we need\n",
    "cleanUp = oriFile[['Allele', 'Peptide', 'Disease', 'Protein', 'Position']]\n",
    "cleanUp['len_peptide'] = cleanUp['Peptide'].str.len()\n",
    "cleanUp.reset_index(inplace=True, drop=True)\n",
    "#cleanUp.rename(columns={'Position':'pos'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "heavy-roberts",
   "metadata": {},
   "source": [
    "## Functions for primer auto-generation and optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lucky-response",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Initialize a dataframe that contains peptide info and a place to store the primers we're about to generate\n",
    "primer_list = cleanUp.copy()\n",
    "primer_list[\"antigen_dna\"] = \"\"\n",
    "primer_list[\"for_primer\"] = \"\"\n",
    "primer_list[\"rev_primer\"] = \"\"\n",
    "primer_list.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "plastic-olympus",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input constant regions for Ii and Linker 2, and the size of the primers to be generated\n",
    "Ii = 'AAAACCTGAGGATGAAGCTGCCTAAACCTCCA'\n",
    "Ii = Ii.lower()\n",
    "L2 = 'GGTGGAGGAAGTTCCGGAGGTGG'\n",
    "L2 = L2.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "consistent-student",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define primer properties: calculate forward and reverse primer lengths, and the variable region length of each primer.\n",
    "def primer_prop(peptide, hyb_len):\n",
    "    frag_len = len(Ii) + len (L2) + len(peptide)*3\n",
    "    for_primer_len = (frag_len + hyb_len)//2\n",
    "    rev_primer_len = frag_len + hyb_len - for_primer_len\n",
    "    primer_for_add = for_primer_len - len(Ii)\n",
    "    primer_rev_add = rev_primer_len - len(L2)\n",
    "    \n",
    "    return primer_for_add, primer_rev_add, frag_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fifty-scholarship",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given a single peptide, reverse translate to human codons with codon optimization (highest frequency in human)\n",
    "def revTrans(peptide):\n",
    "    sense_dna_len = len(peptide)*3\n",
    "    sense_dna = ''\n",
    "    for i in range(len(peptide)):\n",
    "        aa = peptide[i]\n",
    "        ref = sense_5to3_stats[aa]\n",
    "        sense_codon = ref[3][0]\n",
    "        sense_dna += sense_codon\n",
    "        \n",
    "    return sense_dna, sense_dna_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ruled-upset",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a forward and reverse primer based on a sense dna sequence\n",
    "def primerGen(dna):\n",
    "    \n",
    "    # Generating the forward primer\n",
    "    for_primer_str = Ii + dna[0:primer_for_add]\n",
    "    for_primer = Seq(for_primer_str)\n",
    "    \n",
    "    # Generating the reverse primer\n",
    "    rev_primer_str = dna[-primer_rev_add:] + L2\n",
    "    rev_primer = Seq(rev_primer_str)\n",
    "    rev_primer = rev_primer.reverse_complement()\n",
    "    \n",
    "    return for_primer, rev_primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "subjective-cooling",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This is not necessary as we switched to Gibson assembly to insert the antigen fragment\n",
    "## instead of using restriction enzymes. \n",
    "## Checkpoint 1: Check if the peptide DNA fragment (including constant regions) contains restriction enzyme cut sites\n",
    "## input dna as a string\n",
    "def checkpoint1(dna, peptide):\n",
    "    score = 0\n",
    "    sense_dna_Seq = Seq(dna)\n",
    "    Bsu36I_result = Res.Bsu36I.search(sense_dna_Seq)\n",
    "    BspEI_result = Res.BspEI.search(sense_dna_Seq)\n",
    "    if Bsu36I_result != [] or BspEI_result != []:\n",
    "        print(\"For peptide: '%s'\" %peptide)\n",
    "        print(\"Bsu36I or BspEI cut site is observed in the peptide DNA region!\")\n",
    "    else:\n",
    "        score = 1\n",
    "    \n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecological-forest",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Checkpoint 2: check the Tm of the 20 bp hybridization region using Nearest Neighbor thermodynamics\n",
    "\n",
    "# The prediction algorithm used by SnapGene predicts the Tm of the 20 bp hybridiztion region to be \n",
    "# 4-5C higher than the Tm_NN method from BioPyhton. \n",
    "# Updating the NN-table with new parameters used by SnapGene doesn't help much. \n",
    "# Hence, we add 4C to all the melting temperatures predicted by Tm_NN \n",
    "# to check if any of them falls beyond the range 50-69C in which our current PCR recipe works \n",
    "# (i.e. annealing temp).\n",
    "\n",
    "# input for_primer as a seq\n",
    "def checkpoint2(for_primer, peptide, hyb_len):\n",
    "    score = 0\n",
    "    sense_dna_len = len(peptide)*3\n",
    "    hyb_region = for_primer[-hyb_len:]\n",
    "    Tm = mt.Tm_NN(hyb_region) + 4\n",
    "    if Tm < 50 or Tm > 69: \n",
    "        print(\"For peptide: '%s'\" %peptide)\n",
    "        print(\"Tm is %f, beyond the range 50-69!\" %Tm)\n",
    "    else:\n",
    "        print('Tm is %f' %Tm)\n",
    "        score = 1\n",
    "    \n",
    "    return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "subsequent-adoption",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change the melting temperature of the 20bp hybridization region to between 50-69 by changing codons\n",
    "def changeTm(peptide, dna, for_primer, rev_primer):\n",
    "    # test different codon combinations for the aa in the hybridized region, \n",
    "    # excluding the ends of the hybridization region which are inluded in the hairpin structure changes\n",
    "    pos_start = math.ceil((frag_len - len(rev_primer) - len(Ii))/3)\n",
    "    pos_end = math.floor((len(for_primer) - len(Ii))/3)\n",
    "    \n",
    "    aa_pos_change = range(pos_start, pos_end)\n",
    "    start_pos = 0\n",
    "    flag = 'CP2'\n",
    "\n",
    "    # generate complete codon info\n",
    "    aa_to_swap, num_aa = createCodonInfo(aa_pos_change, peptide, start_pos)\n",
    "    \n",
    "    # generate codon combinations and pass through checkpoint 2\n",
    "    # to find hybridization region with a Tm between 50-69C\n",
    "    FOR, REV, dna = genCodonComb(aa_to_swap, num_aa, dna, peptide, flag)\n",
    "    \n",
    "    return FOR, REV, dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adolescent-certification",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test primers with checkpoint3 and optimize primers if failed \n",
    "def checkpoint3(for_primer, rev_primer, peptide, sense_dna, hyb_len):\n",
    "    primer_FOR, primer_REV = '', ''\n",
    "    flag = 'CP23'\n",
    "    score, message = checkpoint3a(for_primer, rev_primer, peptide)\n",
    "    if score != 2:                              # didn't pass CP #3a\n",
    "        print(\"Checkpoint 3a: Check above!\")\n",
    "        primer_FOR, primer_REV = 'Checkpoint 3a', '%s' %message\n",
    "    else:                                       # passed CP #1, 2, and 3a\n",
    "        _, p2, hyb_struc = checkpoint3b_16bphyb(for_primer, rev_primer, peptide, hyb_len)\n",
    "        '''\n",
    "        if p1 != True:\n",
    "            print(\"Checkpoint 3b part 1: Check above!\")\n",
    "        '''\n",
    "        if p2 != True:                          # didn't pass CP #3b\n",
    "            aa_to_swap_for, num_aa_for, aa_to_swap_rev, num_aa_rev = codonToChange(hyb_struc, peptide)\n",
    "            print('aa_to_swap_for')\n",
    "            print(aa_to_swap_for)\n",
    "            print('aa_to_swap_rev')\n",
    "            print(aa_to_swap_rev)\n",
    "            if aa_to_swap_for == [] and aa_to_swap_rev == []:\n",
    "                primer_FOR, primer_REV = 'Less than 16 bp hybridized at 72C!', ''\n",
    "            elif aa_to_swap_for != [] and aa_to_swap_rev != []:\n",
    "                    print('change both primers')\n",
    "                    primer_FOR, primer_REV, sense_dna = genCodonComb(aa_to_swap_for, num_aa_for, \n",
    "                                                                     sense_dna, peptide, flag, \n",
    "                                                                     hairpin_in_both_primers_72C=True)\n",
    "                    print('change rev primer')\n",
    "                    print('sense_dna after removing hairpin in forward primer: %s' %sense_dna)\n",
    "                    if sense_dna == '':\n",
    "                        return\n",
    "                        print('did this return work for if loop: you shouldnt see this')\n",
    "                    else: \n",
    "                        primer_FOR, primer_REV, sense_dna = genCodonComb(aa_to_swap_rev, num_aa_rev, \n",
    "                                                                         sense_dna, peptide, flag) \n",
    "            elif aa_to_swap_for != []:\n",
    "                    primer_FOR, primer_REV, sense_dna = genCodonComb(aa_to_swap_for, num_aa_for, \n",
    "                                                                     sense_dna, peptide, flag)\n",
    "                    print ('sense_dna: %s' %sense_dna)\n",
    "            else:\n",
    "                    print ('sense_dna: %s' %sense_dna)\n",
    "                    primer_FOR, primer_REV, sense_dna = genCodonComb(aa_to_swap_rev, num_aa_rev, \n",
    "                                                                     sense_dna, peptide, flag) \n",
    "                \n",
    "        else:                                   # passed CP #1, 2, 3a and 3b\n",
    "            primer_FOR, primer_REV, sense_dna = for_primer, rev_primer, sense_dna\n",
    "    \n",
    "    return str(primer_FOR), str(primer_REV), str(sense_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "framed-present",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Checkpoint 3a: check using NUPACK at 55C if two primers (1uM) have at least 16bp annealed at the desired location; \n",
    "## AND if the annealed primers are the majority.\n",
    "def checkpoint3a(FOR, REV, peptide):\n",
    "    message = ''\n",
    "    score = 0\n",
    "    '''\n",
    "    define a function\n",
    "    '''\n",
    "    model_55 = npk.Model(material='DNA', celsius = 55)\n",
    "    FOR_str = str(FOR)\n",
    "    REV_str = str(REV)\n",
    "    A = npk.Strand(FOR_str, name = 'A')\n",
    "    B = npk.Strand(REV_str, name = 'B')\n",
    "    t1 = npk.Tube(strands = {A: 7.5e-7, B: 7.5e-7}, \n",
    "                  complexes = npk.SetSpec(max_size = 2),\n",
    "                  name = 't1')\n",
    "    result_55 = npk.tube_analysis(tubes = [t1], model = model_55,\n",
    "                              compute = ['pfunc', 'mfe'])\n",
    "\n",
    "   # Check at 55C if the majority of the 2 primers are annealed together.\n",
    "   # Complex concetration prediction in this version (Nupack 4.0) has a small difference \n",
    "   # compared to the Nupack website prediction results.\n",
    "    for key, value in result_55[t1].complex_concentrations.items():\n",
    "        if key.name == '(A+B)':\n",
    "            print('\\n55C value')\n",
    "            print(value)\n",
    "            if value < 5e-7:\n",
    "                print(\"For peptide: '%s'\" %peptide)\n",
    "                print('Majority of the complexes at 55C are not anneealed!')\n",
    "                print(key.name)\n",
    "                print(value)\n",
    "                message += 'Majority of the complexes at 55C are not anneealed!'\n",
    "            else:\n",
    "                score += 1          \n",
    "\n",
    "    # check if the structure contains >=16 bp hybridization region regardless of hairpin structures \n",
    "    '''\n",
    "    Experimental results: 16-20 bp hybridization works. \n",
    "    '''\n",
    "    AB_55 = result_55['(A+B)']\n",
    "    AB_55_struc = str(AB_55.mfe[0].structure)\n",
    "    print('55C struc')\n",
    "    print(AB_55_struc)\n",
    "    for_annealed = '(((((((((((((((('          # rev_annealed = ')))))))))))))))))'\n",
    "    if for_annealed not in AB_55_struc[frag_len-len(REV):(len(FOR)+1)]:\n",
    "        print(\"For peptide: '%s'\" %peptide)\n",
    "        print('Less than 16 bp annealed!')\n",
    "        message += 'Less than 16 bp annealed!'\n",
    "    else:\n",
    "        score += 1\n",
    "    \n",
    "    return score, message"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "photographic-heart",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Checkpoint 3b: check using NUPACK at 72C if two primers (1uM) have at least 16 bp \n",
    "## annealed at the desired location; \n",
    "## AND if hairpin structure exists that might hinder the extension process.\n",
    "def checkpoint3b_16bphyb(FOR, REV, peptide, hyb_len, check_for_primer_first = False):\n",
    "    score = 0\n",
    "    part1 = False\n",
    "    part2 = False\n",
    "    model_72 = npk.Model(material='DNA', celsius = 72)\n",
    "    FOR_str = str(FOR)\n",
    "    REV_str = str(REV)\n",
    "    A = npk.Strand(FOR_str, name = 'A')\n",
    "    B = npk.Strand(REV_str, name = 'B')\n",
    "    t1 = npk.Tube(strands = {A: 7.5e-7, B: 7.5e-7}, \n",
    "                  complexes = npk.SetSpec(max_size = 2),\n",
    "                  name = 't1')\n",
    "    result_72 = npk.tube_analysis(tubes = [t1], model = model_72,\n",
    "                              compute = ['pfunc', 'mfe'])\n",
    "    \n",
    "    # check using NUPACK at 72C to determine if a minimum of 16 bp is hybridized\n",
    "    # check using NUPACK at 72C if a hairpin structure forms which might hinder the extension process\n",
    "    for_con = len(FOR_str) - hyb_len\n",
    "    rev_con = len(REV_str) - hyb_len\n",
    "    \n",
    "    # initialize the constants for primer hybridization structures\n",
    "    for_1 = for_con + 1\n",
    "    for_2 = for_con + 2\n",
    "    for_3 = for_con + 3\n",
    "    for_4 = for_con + 4\n",
    "    \n",
    "    rev_1 = rev_con + 1\n",
    "    rev_2 = rev_con + 2\n",
    "    rev_3 = rev_con + 3\n",
    "    rev_4 = rev_con + 4\n",
    "    \n",
    "    hyb_1 = hyb_len - 1\n",
    "    hyb_2 = hyb_len - 2\n",
    "    hyb_3 = hyb_len - 3\n",
    "    hyb_4 = hyb_len - 4\n",
    "    \n",
    "    perfect_struc = npk.Structure(f'.{for_con}({hyb_len}+.{rev_con}){hyb_len}')\n",
    "    # 1 bp unhybridized\n",
    "    unhyb_1bp_1 = npk.Structure(f'.{for_1}({hyb_1}+.{rev_con}){hyb_1}.1')\n",
    "    unhyb_1bp_2 = npk.Structure(f'.{for_con}({hyb_1}.1+.{rev_1}){hyb_1}')\n",
    "    # 2 bp unhybridized\n",
    "    unhyb_2bp_1 = npk.Structure(f'.{for_2}({hyb_2}+.{rev_con}){hyb_2}.2')\n",
    "    unhyb_2bp_2 = npk.Structure(f'.{for_con}({hyb_2}.2+.{rev_2}){hyb_2}')\n",
    "    unhyb_2bp_3 = npk.Structure(f'.{for_1}({hyb_2}.1+.{rev_1}){hyb_2}.1')\n",
    "    # 3 bp unhybridized\n",
    "    unhyb_3bp_1 = npk.Structure(f'.{for_3}({hyb_3}+.{rev_con}){hyb_3}.3')\n",
    "    unhyb_3bp_2 = npk.Structure(f'.{for_con}({hyb_3}.3+.{rev_3}){hyb_3}')\n",
    "    unhyb_3bp_3 = npk.Structure(f'.{for_2}({hyb_3}.1+.{rev_1}){hyb_3}.2')\n",
    "    unhyb_3bp_4 = npk.Structure(f'.{for_1}({hyb_3}.2+.{rev_2}){hyb_3}.1')\n",
    "    # 4 bp unhybridized\n",
    "    unhyb_4bp_1 = npk.Structure(f'.{for_4}({hyb_4}+.{rev_con}){hyb_4}.4')\n",
    "    unhyb_4bp_2 = npk.Structure(f'.{for_con}({hyb_4}.4+.{rev_4}){hyb_4}')\n",
    "    unhyb_4bp_3 = npk.Structure(f'.{for_3}({hyb_4}.1+.{rev_1}){hyb_4}.3')\n",
    "    unhyb_4bp_4 = npk.Structure(f'.{for_1}({hyb_4}.3+.{rev_3}){hyb_4}.1')\n",
    "    unhyb_4bp_5 = npk.Structure(f'.{for_2}({hyb_4}.2+.{rev_2}){hyb_4}.2')\n",
    "             \n",
    "    ideal_struc_list = [perfect_struc, \n",
    "                        unhyb_1bp_1, unhyb_1bp_2, \n",
    "                        unhyb_2bp_1, unhyb_2bp_2, unhyb_2bp_3,\n",
    "                        unhyb_3bp_1, unhyb_3bp_2, unhyb_3bp_3, unhyb_3bp_4,\n",
    "                        unhyb_4bp_1, unhyb_4bp_2, unhyb_4bp_3, unhyb_4bp_4, unhyb_4bp_5]\n",
    "    # 16-mer peptide, max 17 bp hybridized, min 16 bp hybridized\n",
    "    ideal_struc_list_16mer = [perfect_struc, \n",
    "                              unhyb_1bp_1, unhyb_1bp_2]\n",
    "    \n",
    "    AB_72 = result_72['(A+B)']\n",
    "    AB_72_struc = AB_72.mfe[0].structure\n",
    "    print('72C struc')\n",
    "    print(AB_72_struc)\n",
    "    print('FOR: %s' %FOR_str)\n",
    "    print('REV: %s' %REV_str)\n",
    "      \n",
    "    if check_for_primer_first == True:\n",
    "        if str(npk.Structure(f'.{for_con}')) in str(AB_72_struc)[:for_con]:\n",
    "            score += 1\n",
    "            return score, part2, AB_72_struc\n",
    "        else:\n",
    "            return score, part2, AB_72_struc\n",
    "    \n",
    "    if len(peptide) == 16:\n",
    "        ideal_struc = ideal_struc_list_16mer\n",
    "    else:\n",
    "        ideal_struc = ideal_struc_list\n",
    "        \n",
    "    if AB_72_struc not in ideal_struc:\n",
    "        print('Hairpin structure or < 16 bp unhybridized regions might exist in primer at 72C!')\n",
    "    else:\n",
    "        part2 = True\n",
    "        score += 1\n",
    "    \n",
    "    return score, part2, AB_72_struc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "broad-brush",
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify the position of the amino acid in the peptide list, whose codon will be changed\n",
    "def codonPos(aa_num, dna_struc_slice):\n",
    "    aa_pos_change = []\n",
    "    for i in range(aa_num):\n",
    "        pos_end = (i+1)*3 - 1\n",
    "        codon = dna_struc_slice[pos_end-2:pos_end+1]\n",
    "        if ')' in codon or '(' in codon:\n",
    "            aa_pos_change.append(i)\n",
    " \n",
    "    return aa_pos_change"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "connected-headline",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create an array of peptides whose codon will be changed, info includes: \n",
    "# name of aa, candidate codons, swap_index for generating different combinations\n",
    "# of codons, and the codon location in the sense dna that will be changed\n",
    "def createCodonInfo(aa_pos_change, peptide, start_pos):\n",
    "    aa_to_swap = []\n",
    "    for i in aa_pos_change:\n",
    "        aa = {}\n",
    "        aa['name'] = peptide[i + start_pos]\n",
    "        aa['candidates'] = sense_5to3_stats[aa['name']][3] # I'm allowing the _reverse primer: first bp of the first codon \n",
    "        aa['swap_index'] = 0                               # OR forward primer: last bp of last codon_ to be changed as well, \n",
    "        aa['index_sense_dna'] = (i+start_pos)*3            # which is in hybridization region, \n",
    "        aa_to_swap.append(aa)                              # this will not change the result in Checkpoint #2 (55C annealing).\n",
    "    num_aa = len(aa_to_swap) \n",
    "    \n",
    "    return aa_to_swap, num_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "julian-terror",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather info needed to remove hairpin in 72C primer struc\n",
    "def codonToChange(struc_72C, peptide):\n",
    "    # Identify editable positions that are outside the constant region and hybridization region\n",
    "    struc_72C = str(struc_72C) \n",
    "    print('check struc_72C')\n",
    "    print(struc_72C)\n",
    "    stop_for = frag_len - len(rev_primer)\n",
    "    struc_for = struc_72C[len(Ii):stop_for]\n",
    "    aa_num_for = math.ceil(len(struc_for)/3)\n",
    "    \n",
    "    start_rev = len(for_primer) + 1 + len(L2)\n",
    "    struc_rev = struc_72C[start_rev:-hyb_len]\n",
    "    rev_pos = struc_rev[::-1]\n",
    "    aa_num_rev = math.ceil(len(struc_rev)/3)\n",
    "    # allowing to change the end of the hybridization sites\n",
    "    if (3-len(struc_rev)%3) == 1:\n",
    "        rev_pos_added = '.' + rev_pos\n",
    "    elif (3-len(struc_rev)%3) == 2:\n",
    "        rev_pos_added = '..' + rev_pos\n",
    "        \n",
    "    # Define the starting position of the corresponding amino acid in the peptide\n",
    "    start_pos_for = 0\n",
    "    start_pos_rev = math.floor((len(for_primer) - len(Ii))/3)\n",
    "\n",
    "    # codon to be changed in forward/reverse primer\n",
    "    aa_pos_change_for = codonPos(aa_num_for, struc_for)\n",
    "    aa_pos_change_rev = codonPos(aa_num_rev, rev_pos_added)\n",
    "\n",
    "    # information needed for changing codons   \n",
    "    aa_to_swap_for, num_aa_for = createCodonInfo(aa_pos_change_for, peptide, start_pos_for)\n",
    "    aa_to_swap_rev, num_aa_rev = createCodonInfo(aa_pos_change_rev, peptide, start_pos_rev)\n",
    "    \n",
    "    return aa_to_swap_for, num_aa_for, aa_to_swap_rev, num_aa_rev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divided-shannon",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate different codon combinations on a selected set amino acids\n",
    "def genCodonComb(aa_to_swap, num_aa, dna, peptide, flag, hairpin_in_both_primers_72C = False):\n",
    "    total_iter = 1\n",
    "    \n",
    "    for i in range(num_aa):\n",
    "        total_iter *= len(aa_to_swap[i]['candidates'])\n",
    "    \n",
    "    print('here')\n",
    "    print('total_iter %d' %total_iter)\n",
    "    \n",
    "    for i in range(total_iter):\n",
    "         # generate new seq\n",
    "        for j in range(num_aa):\n",
    "            amino_acid = aa_to_swap[j]\n",
    "            pos = amino_acid['index_sense_dna']\n",
    "            swap = amino_acid['swap_index']\n",
    "            dna = dna[:pos] + amino_acid['candidates'][swap] + dna[pos+3:]\n",
    "        \n",
    "        # generate new forward and reverse primers using the new sense DNA sequence\n",
    "        FOR, REV = primerGen(dna)\n",
    "        print('FOR_test: %s' %FOR)\n",
    "        print('REV_test: %s' %REV)\n",
    "        \n",
    "        # determine which checkpoints to check\n",
    "        if flag == 'CP23':\n",
    "            message = 'Unable to find optimized primers!'\n",
    "            # check through Checkpoint 2, 3a, and 3b\n",
    "            score3a, _ = checkpoint3a(FOR, REV, peptide)\n",
    "            score3b, _, _ = checkpoint3b_16bphyb(FOR, REV, peptide, hyb_len)\n",
    "            if hairpin_in_both_primers_72C == True:\n",
    "                print('here-1')\n",
    "                score3b, _, _ = checkpoint3b_16bphyb(FOR, REV, peptide, hyb_len, check_for_primer_first = True)\n",
    "                if checkpoint2(FOR, peptide, hyb_len) + score3a + score3b == 4:\n",
    "                    print('here-2')\n",
    "                    return FOR, REV, dna\n",
    "                    break\n",
    "            if checkpoint2(FOR, peptide, hyb_len) + score3a + score3b == 4:\n",
    "                print('Sense_dna: %s' %dna)\n",
    "                print('Forward primer: %s' %FOR)\n",
    "                print('Reverse primer: %s' %REV)\n",
    "                return FOR, REV, dna\n",
    "                break\n",
    "        elif flag == 'CP2':\n",
    "            message = 'Unable to make Tm in the 50-69C range!'\n",
    "            # check through checkpoint 2\n",
    "            if checkpoint2(FOR, peptide, hyb_len) == 1:\n",
    "                print('Checkpoint 2 passed.')\n",
    "                return FOR, REV, dna\n",
    "                break\n",
    "        \n",
    "        # increment the swap index\n",
    "        for j in range(num_aa):\n",
    "            aa_to_swap[j]['swap_index'] += 1\n",
    "            if aa_to_swap[j]['swap_index'] == len(aa_to_swap[j]['candidates']):\n",
    "                aa_to_swap[j]['swap_index'] = 0\n",
    "            else:\n",
    "                break\n",
    "        \n",
    "        if i == total_iter - 1:\n",
    "            print(\"%s\" %message)\n",
    "            FOR, REV, dna = '%s' %message, '', ''\n",
    "            return FOR, REV, dna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5d155f-393c-4f49-b48b-5492c837dd64",
   "metadata": {},
   "source": [
    "## Generate and optimize primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "short-resistance",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate forward and reverse primers for each peptide; \n",
    "# Check if the primers satisfy the following criteria: checkpoint 1, 2 and 3.\n",
    "\n",
    "for index in primer_list.index:\n",
    "    \n",
    "    print('\\nPos %d' %index)\n",
    "        \n",
    "    peptide = primer_list.loc[index, 'Peptide']\n",
    "    if len(peptide) == 16:\n",
    "        hyb_len = 17\n",
    "    else:\n",
    "        hyb_len = 20\n",
    "    sense_dna, sense_dna_len = revTrans(peptide)\n",
    "    total_dna = Ii[-10:] + sense_dna + L2[:10]\n",
    "    \n",
    "    # overall workflow\n",
    "\n",
    "    primer_for_add, primer_rev_add, frag_len = primer_prop(peptide, hyb_len)\n",
    "    for_primer, rev_primer = primerGen(sense_dna)\n",
    "    \n",
    "    # If using Gibson assembly, then checkpoint 1 is not necessary.\n",
    "    # CP #1 is only needed if double digestion using restriction enzymes are used for inserting the antigen fragments \n",
    "    # into the pcDNA plasmid.\n",
    "    '''\n",
    "    if checkpoint1(total_dna, peptide) != 1:        # didn't pass CP #1\n",
    "        print(\"Checkpoint 1: Check above!\")\n",
    "        primer_FOR, primer_REV = 'Checkpoint 1', ''\n",
    "    '''\n",
    "    if checkpoint2(for_primer, peptide, hyb_len) != 1:     # passed CP #1, but didn't pass CP #2\n",
    "        for_primer, rev_primer, sense_dna = changeTm(peptide, sense_dna, for_primer, rev_primer)\n",
    "        if for_primer != 'Unable to make Tm in the 50-69C range!':\n",
    "            primer_FOR, primer_REV, sense_dna = checkpoint3(for_primer, rev_primer, \n",
    "                                                            peptide, sense_dna, hyb_len)        \n",
    "        else:\n",
    "            primer_FOR, primer_REV = 'Unable to make Tm in the 50-69C range!', ''\n",
    "    else:                                           # passed CP #1 and 2\n",
    "        print(\"for_primer %s\" %for_primer)\n",
    "        print(\"rev_primer %s\" %rev_primer)\n",
    "        primer_FOR, primer_REV, sense_dna = checkpoint3(for_primer, rev_primer, \n",
    "                                                        peptide, sense_dna, hyb_len)\n",
    "    \n",
    "    print(sense_dna)\n",
    "    primer_list.loc[index, 'antigen_dna'] = str(sense_dna)\n",
    "    primer_list.loc[index, 'for_primer'] = str(primer_FOR)\n",
    "    primer_list.loc[index, 'rev_primer'] = str(primer_REV)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09847dca-cacc-4d50-847b-95ad2fdc09de",
   "metadata": {},
   "source": [
    "## Ouptut the optimized primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "enabling-navigator",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "primer_list['ID'] = primer_list.index + 1\n",
    "primer_list['for_primer_name'] = 'DR_DP_FOR_' + primer_list['ID'].astype(str)\n",
    "primer_list['rev_primer_name'] = 'DR_DP_REV_' + primer_list['ID'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6391053-289b-4fa9-989b-57a0842f1970",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = input(\"Please enter the file path for the output file: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aboriginal-determination",
   "metadata": {},
   "outputs": [],
   "source": [
    "# output the primers generated. Need to manually check the ones with error messages.\n",
    "primer_list.to_excel(output_path)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
